### PR TITLE
[5.6] Refactor & Improve Scheduler Events

### PR DESCRIPTION
**Pull request by laurencei**
I've been doing a lot of digging around the scheduler lately. I started re-writing it for my package to give better monitoring - but I think what I've ended up with is pretty cool and would be better in the framework itself.

This PR gives the following changes to schedulers:

- Closures/Callbacks can now be run in background!
- Closures/Callbacks can now have their output captured (and either emailed or stored)!
- Both Foreground and Background events follow a similar flow (more DRY)
- Background events no longer need `schedule:finish`
- Command and Closure events follow a similar flow (more DRY)
- Better mutex handling should result in less 'lock' scenarios when a command fails (all paths now lead to one place - so we handle `register_shutdown_function` there).

Essentially what occurs is background events are now "bounced" off a `scheduler:background` command in the background, but all that command does is re-run the event in the "foreground" (but on a different process). I drew a diagram to try and illustrate the change in flow:

flow

This allows us to DRY up the files, and effectively have a single flow for all events. It also means all events can be handled in a single flow, and all of them get the same benefits.

My only question is; is it ok to use `ob_start()` etc to capture the callback output (to allow for emailing etc)? It is working (I tested it) - but I wanted to check if any side effects I'm not considering? That feature can be removed if it clashes with the framework somehow.

I've done lots of actual testing of this code on my project, trying all sorts of combinations of commands, background, mutexs etc - and it all seems to work flawlessly. But would be great if a couple of more people could test, especially if you have funky schedules or commands etc.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Where is anything run in the background?

**Comment by laurencei** [comment]
Anything set to `runInBackground` is now sent directly to `schedule:background` as a background process with the mutex id.

`schedule:background` (now running on it's "own" process - i.e. background) - re-runs the mutex and forces it to run in the foreground (buts its really in the background)

**Comment by laurencei** [comment]
Here is a demo of it working:

example

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php [hunk 1]
```diff
@@ -54,8 +54,24 @@ public function __construct(Schedule $schedule)
...
+    }
+
+    /**
+     * Find the event that matches the id.
+     *
+     * @return mixd
+     */
+    protected function findEventMutex()
+    {
+        return collect($this->schedule->events())->filter(function ($value) {
```
Comment (lucasmichot) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: Isn't `\Illuminate\Support\Arr::first` just enought for that ?

File: src/Illuminate/Console/Scheduling/CallbackEvent.php [hunk 2]
```diff
@@ -119,20 +91,39 @@ public function withoutOverlapping($expiresAt = 1440)
...
+        }
+
+        return parent::runInBackground();
+    }
+
+    /**
+     * Store any captured output.
+     *
+     * @param  string  $output
+     * @return mixed
```
Comment (lucasmichot) [inline src/Illuminate/Console/Scheduling/CallbackEvent.php]: `@return string` instead ?

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php (hunk 1 above)
Comment (laurencei) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: mmmm... I dont think so? You need to filter all events against the mutex, which is a command to check.

This is similar to the command Taylor already had for `schedule:background` in 5.4 - so I was trying to keep the logic similar where possible to minimize changes.

File: src/Illuminate/Console/Scheduling/CallbackEvent.php (hunk 2 above)
Comment (laurencei) [inline src/Illuminate/Console/Scheduling/CallbackEvent.php]: yep - thanks

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php (hunk 1 above)
Comment (lucasmichot) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: Actually here you filter your collection, then you return the first item of this filtered collection.
With Arr::first, you return the first item that match the condition in the closure.

This will achieve the same thing, but much faster:
```php
return Arr::first($this->schedule->events(), (function (Event $value) {
    return $value->mutexName() === $this->argument('id');
});
```

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php [hunk 3]
```diff
@@ -54,8 +54,24 @@ public function __construct(Schedule $schedule)
...
+            return;
+        }
+
+        $event->runCommandInForeground($this->laravel);
+    }
+
+    /**
+     * Find the event that matches the id.
+     *
+     * @return mixd
```
Comment (lucasmichot) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: `@return \Illuminate\Console\Scheduling\Event|null`

File: src/Illuminate/Console/Scheduling/CallbackEvent.php [hunk 4]
```diff
@@ -119,20 +91,39 @@ public function withoutOverlapping($expiresAt = 1440)
...
     {
-        return 'framework/schedule-'.sha1($this->description);
+        return 'schedule-mutex-'.sha1($this->expression.$this->description);
+    }
+
+    /**
+     * Do not allow the event to run in background without a name.
+     *
+     * @return $this
+     */
```
Comment (lucasmichot) [inline src/Illuminate/Console/Scheduling/CallbackEvent.php]: `@throws \LogicException` maybe ?

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php (hunk 1 above)
Comment (laurencei) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: ok

File: src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php (hunk 3 above)
Comment (laurencei) [inline src/Illuminate/Console/Scheduling/ScheduleBackgroundCommand.php]: done

File: src/Illuminate/Console/Scheduling/CallbackEvent.php (hunk 4 above)
Comment (laurencei) [inline src/Illuminate/Console/Scheduling/CallbackEvent.php]: done

**Comment by taylorotwell** [comment]
I dunno, I think I'm going to hold off on this for now. If there is any bugs in the current system would rather address them in a less radical way.

**Comment by taylorotwell** [comment]
Meh, I'll read through it one more time.

**Comment by laurencei** [comment]
Hi Taylor,

There are no bugs in the current system - just "limitations".

It's not particularly radical. It is basically taking your own concept of using `schedule:finish` to apply the after callbacks - and just taking it two steps further, to run the _whole_ event instead.

By doing this, it then allows you to then include the callbacks in the background method etc.

**Comment by taylorotwell** [comment]
Can you describe the need for the `ob_end_flush`?

**Comment by taylorotwell** [comment]
Using the "Log::error" calls is a no-no inside a component. You will just have to throw the exception.

**Comment by laurencei** [comment]
> Can you describe the need for the ob_end_flush?

This is only used in the callback section. It allows for output to be captured and stored. This gets around the limitation that we are not running a process (where the operating system will push the output to a file for us).

If output is **not** being stored, we just just run the callback and return:

    if ($this->outputNotBeingCaptured()) {
       return $container->call($this->callback, $this->parameters);
    }

Otherwise, we turn on `op_start()` to start capturing the output. The output is captured via `ob_get_contents()` and stored in a variable.

We then run `op_end_flush` to end the output buffer *and* dump the output to the screen - because someone might want to see if (if they were running the command manually etc).

Basically the whole `op_start()` etc stuff is optional - as it is only to allow the callback outputs to be optionally captured.

**Comment by andrewrcollins** [comment]
There is a minor bug with `before()` callbacks not be called for `CallbackEvent` events:

See #20804

**Comment by laurencei** [comment]
Yes. This PR will fix that problem. 👍

**Comment by GrahamCampbell** [comment]
Would be nice to see this in 5.6. Maybe too late for 5.5 now, since fixing that ticket is technically BC breaking?

**Comment by laurencei** [comment]
To be honest - I dont think it is a breaking change - it should be 100% compatible with all existing code.

But I do agree that it is enough change in the overall flow in the code that it probably makes more sense to target 5.6 now.

**Comment by laurencei** [comment]
@taylorotwell - is this something you are still interested in?

I only ask, because there are now merge conflicts - so do you want me to fix this PR to remove the conflict?

Let me know. Thanks.

**Comment by taylorotwell** [comment]
I don't think I'm going to go forward with this right now.
